node/request.js

```diff
@@ -226,32 +228,50 @@ TChannelRequest.prototype.resend = function resend() {
         opts[key] = self.options[key];
     }
     opts.timeout = self.timeout - self.elapsed;
-    var outReq = peer.request(opts);
-    self.outReqs.push(outReq);
 
-    if (self.outReqs.length === 1) {
-        self.channel.outboundCallsSentStat.increment(1, {
-            'target-service': outReq.serviceName,
-            'service': outReq.headers.cn,
-            // TODO should always be buffer
-            'target-endpoint': String(self.arg1)
+    self.services.onRequest(self);
```

Comment:
For the purposes of back-pressure, services.onRequest should probably be called as soon as the request is queued, and only once for each corresponding error or response case. Having this in resend might actually be too late since resend gets called once for each retry. To maintain the invariant that there’s one error or response for each request, we must either onError before scheduling a retry, or onRequest once per send, not resend. I’m favoring the latter since a request that is scheduled for retry should count toward the max pending requests limit.